File: linux/device/go/internal/stream/session.go
// SPDX-License-Identifier: AGPL-3.0-only

package stream

import (
	"errors"
	"log/slog"
	"sync"
	"time"
)

// Transport is a message-oriented binary carrier; each ReadMessage/WriteMessage moves one whole HelixStream frame.
type Transport interface {
	ReadMessage() ([]byte, error)
	WriteMessage([]byte) error
	Close() error
}

// Config tunes a Session.
type Config struct {
	// Client selects the stream-id parity so both peers can open without collision: client odd, server even.
	Client bool
	// InitialWindow is the per-stream receive window in bytes (default 256 KiB).
	InitialWindow int
	// MaxChunk is the largest DATA payload per frame (default 32 KiB).
	MaxChunk int
	// Logger is optional.
	Logger *slog.Logger
}

// ErrSessionClosed is returned once the session is torn down.
var ErrSessionClosed = errors.New("stream session closed")

// Session multiplexes Streams over one Transport.
type Session struct {
	t   Transport
	log *slog.Logger

	initialWindow   int
	maxChunk        int
	creditThreshold int

	writeMu sync.Mutex

	mu      sync.Mutex
	streams map[uint32]*Stream
	nextID  uint32

	accept    chan *Stream
	closeOnce sync.Once
	closed    chan struct{}
	closeErr  error
}

// NewSession wraps a Transport and starts serving frames.
func NewSession(t Transport, cfg Config) *Session {
	if cfg.InitialWindow <= 0 {
		cfg.InitialWindow = 256 * 1024
	}
	if cfg.MaxChunk <= 0 {
		cfg.MaxChunk = 32 * 1024
	}
	if cfg.Logger == nil {
		cfg.Logger = slog.Default()
	}
	firstID := uint32(2)
	if cfg.Client {
		firstID = 1
	}
	s := &Session{
		t:               t,
		log:             cfg.Logger,
		initialWindow:   cfg.InitialWindow,
		maxChunk:        cfg.MaxChunk,
		creditThreshold: cfg.InitialWindow / 2,
		streams:         make(map[uint32]*Stream),
		nextID:          firstID,
		accept:          make(chan *Stream, 64),
		closed:          make(chan struct{}),
	}
	go s.readLoop()
	go s.pingLoop()
	return s
}

// Open starts a new local stream; meta is an app-defined open payload (may nil).
func (s *Session) Open(meta []byte) (*Stream, error) {
	s.mu.Lock()
	select {
	case <-s.closed:
		s.mu.Unlock()
		return nil, ErrSessionClosed
	default:
	}
	id := s.nextID
	s.nextID += 2
	st := newStream(s, id, meta)
	s.streams[id] = st
	s.mu.Unlock()

	if err := s.writeFrame(fOpen, id, meta); err != nil {
		s.removeStream(id)
		return nil, err
	}
	return st, nil
}

// Accept returns the next peer-initiated stream, or an error once closed.
func (s *Session) Accept() (*Stream, error) {
	select {
	case st := <-s.accept:
		return st, nil
	case <-s.closed:
		return nil, s.closeErr
	}
}

// Close tears down the session and all streams.
func (s *Session) Close() error {
	s.closeWithErr(ErrSessionClosed)
	return nil
}

// Done is closed when the session ends.
func (s *Session) Done() <-chan struct{} { return s.closed }

// StreamCount returns the number of streams currently open on the session.
func (s *Session) StreamCount() int {
	s.mu.Lock()
	defer s.mu.Unlock()
	return len(s.streams)
}

func (s *Session) writeFrame(t frameType, id uint32, payload []byte) error {
	frame := encodeFrame(t, id, payload)
	s.writeMu.Lock()
	err := s.t.WriteMessage(frame)
	s.writeMu.Unlock()
	if err != nil {
		s.closeWithErr(err)
	}
	return err
}

func (s *Session) removeStream(id uint32) {
	s.mu.Lock()
	delete(s.streams, id)
	s.mu.Unlock()
}

func (s *Session) getStream(id uint32) *Stream {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.streams[id]
}

func (s *Session) readLoop() {
	for {
		msg, err := s.t.ReadMessage()
		if err != nil {
			s.closeWithErr(err)
			return
		}
		t, id, payload, ok := decodeFrame(msg)
		if !ok {
			continue
		}
		s.dispatch(t, id, payload)
	}
}

func (s *Session) dispatch(t frameType, id uint32, payload []byte) {
	switch t {
	case fPing:
		_ = s.writeFrame(fPong, 0, nil)
	case fPong:
	case fOpen:
		st := newStream(s, id, append([]byte(nil), payload...))
		s.mu.Lock()
		s.streams[id] = st
		s.mu.Unlock()
		select {
		case s.accept <- st:
		case <-s.closed:
		}
	case fData:
		if st := s.getStream(id); st != nil {
			st.onData(payload)
		}
	case fEnd:
		if st := s.getStream(id); st != nil {
			st.onEnd()
		}
	case fReset:
		if st := s.getStream(id); st != nil {
			st.teardown(errors.New("peer reset"))
			s.removeStream(id)
		}
	case fSignal:
		if st := s.getStream(id); st != nil {
			st.onSignal(payload)
		}
	case fCredit:
		if st := s.getStream(id); st != nil {
			st.onCredit(decodeCredit(payload))
		}
	}
}

func (s *Session) pingLoop() {
	ticker := time.NewTicker(20 * time.Second)
	defer ticker.Stop()
	for {
		select {
		case <-s.closed:
			return
		case <-ticker.C:
			_ = s.writeFrame(fPing, 0, nil)
		}
	}
}

func (s *Session) closeWithErr(err error) {
	s.closeOnce.Do(func() {
		s.closeErr = err
		close(s.closed)
		s.mu.Lock()
		streams := make([]*Stream, 0, len(s.streams))
		for _, st := range s.streams {
			streams = append(streams, st)
		}
		s.streams = make(map[uint32]*Stream)
		s.mu.Unlock()
		for _, st := range streams {
			st.teardown(err)
		}
		_ = s.t.Close()
	})
}

File: web/packages/protocol/src/stream/session.ts
import { decodeCredit, decodeFrame, encodeFrame, FrameType } from './frame';
import { HelixStream, type StreamHost } from './stream';

import type { StreamTransport } from './transport';

/** Defaults; must match the Go stream.Session (internal/stream/session.go). */
const DEFAULT_INITIAL_WINDOW_BYTES = 262_144; // 256 KiB
const DEFAULT_MAX_CHUNK_BYTES = 32_768; // 32 KiB
const DEFAULT_PING_INTERVAL_MS = 20_000;
/** Grant credit once the peer has consumed half the window. */
const CREDIT_THRESHOLD_DIVISOR = 2;
/** client => odd stream ids, server => even, so neither peer collides. */
const FIRST_CLIENT_STREAM_ID = 1;
const FIRST_SERVER_STREAM_ID = 2;

export type SessionOptions = Readonly<{
  /** client => odd stream ids, server => even, so both peers can open. */
  client: boolean;
  /** Per-stream receive window in bytes (default 256 KiB). */
  initialWindow?: number;
  /** Largest DATA payload per frame (default 32 KiB). */
  maxChunk?: number;
  /** Called for each peer-initiated stream. */
  onStream?: (stream: HelixStream) => void;
  /** Keepalive ping interval in ms (default 20000; 0 disables). */
  pingIntervalMs?: number;
}>;

/** Multiplexes many HelixStreams over one binary transport; byte-compatible with Go stream.Session. */
export class HelixStreamSession implements StreamHost {
  readonly creditThreshold: number;
  readonly maxChunk: number;

  onStream?: (stream: HelixStream) => void;
  onClose?: (err?: Error) => void;

  private readonly initialWindow: number;
  private readonly streams = new Map<number, HelixStream>();
  private nextId: number;
  private closed = false;
  private readonly pingTimer?: ReturnType<typeof setInterval>;

  constructor(
    private readonly transport: StreamTransport,
    opts: SessionOptions,
  ) {
    this.initialWindow = opts.initialWindow ?? DEFAULT_INITIAL_WINDOW_BYTES;
    this.maxChunk = opts.maxChunk ?? DEFAULT_MAX_CHUNK_BYTES;
    this.creditThreshold = Math.floor(this.initialWindow / CREDIT_THRESHOLD_DIVISOR);
    this.nextId = opts.client ? FIRST_CLIENT_STREAM_ID : FIRST_SERVER_STREAM_ID;
    this.onStream = opts.onStream;

    transport.onMessage = (data) => {
      this.dispatch(data);
    };
    transport.onClose = (err) => {
      this.closeWithErr(err);
    };

    const interval = opts.pingIntervalMs ?? DEFAULT_PING_INTERVAL_MS;
    if (interval > 0) {
      this.pingTimer = setInterval(() => {
        this.sendFrame(FrameType.Ping, 0);
      }, interval);
    }
  }

  /** Open a new local stream; meta is an app-defined open payload. */
  open(meta?: Uint8Array): HelixStream {
    if (this.closed) {
      throw new Error('stream session closed');
    }
    const id = this.nextId;
    this.nextId += 2;
    const stream = new HelixStream(id, meta ?? new Uint8Array(0), this, this.initialWindow);
    this.streams.set(id, stream);
    this.sendFrame(FrameType.Open, id, meta);
    return stream;
  }

  close(): void {
    this.closeWithErr(undefined);
  }

  sendFrame(type: FrameType, streamId: number, payload?: Uint8Array): void {
    if (this.closed) {
      return;
    }
    try {
      this.transport.send(encodeFrame(type, streamId, payload));
    } catch (err) {
      this.closeWithErr(err instanceof Error ? err : new Error(String(err)));
    }
  }

  removeStream(id: number): void {
    this.streams.delete(id);
  }

  private dispatch(data: Uint8Array): void {
    const frame = decodeFrame(data);
    if (frame === null) {
      return;
    }
    const { type, streamId, payload } = frame;
    switch (type) {
      case FrameType.Ping:
        this.sendFrame(FrameType.Pong, 0);
        return;
      case FrameType.Pong:
        return;
      case FrameType.Open: {
        const stream = new HelixStream(streamId, payload.slice(), this, this.initialWindow);
        this.streams.set(streamId, stream);
        this.onStream?.(stream);
        return;
      }
      case FrameType.Data:
        this.streams.get(streamId)?._onData(payload);
        return;
      case FrameType.End:
        this.streams.get(streamId)?._onEnd();
        return;
      case FrameType.Reset: {
        const s = this.streams.get(streamId);
        if (s !== undefined) {
          s.teardown(new Error('peer reset'));
          this.streams.delete(streamId);
        }
        return;
      }
      case FrameType.Signal:
        this.streams.get(streamId)?._onSignal(payload);
        return;
      case FrameType.Credit:
        this.streams.get(streamId)?._onCredit(decodeCredit(payload));
        return;
      default:
        return;
    }
  }

  private closeWithErr(err?: Error): void {
    if (this.closed) {
      return;
    }
    this.closed = true;
    if (this.pingTimer !== undefined) {
      clearInterval(this.pingTimer);
    }
    for (const stream of this.streams.values()) {
      stream.teardown(err);
    }
    this.streams.clear();
    try {
      this.transport.close();
    } catch {
      /* already closed */
    }
    this.onClose?.(err);
  }
}
